l2geth/rollup/sync_service.go

```diff
@@ -851,32 +848,47 @@ func (s *SyncService) verifyFee(tx *types.Transaction) error {
 	if err != nil {
 		return err
 	}
-	// Calculate the fee based on decoded L2 gas limit
-	gas := new(big.Int).SetUint64(tx.Gas())
-	l2GasLimit := fees.DecodeL2GasLimit(gas)
 
-	// Only count the calldata here as the overhead of the fully encoded
-	// RLP transaction is handled inside of EncodeL2GasLimit
-	expectedTxGasLimit := fees.EncodeTxGasLimit(tx.Data(), l1GasPrice, l2GasLimit, l2GasPrice)
-
-	// This should only happen if the unscaled transaction fee is greater than 18.44 ETH
-	if !expectedTxGasLimit.IsUint64() {
-		return fmt.Errorf("fee overflow: %s", expectedTxGasLimit.String())
+	// Check that the user has enough funds to cover the fee
+	// (l1GasPrice + overhead * calldata cost) + (l2GasPrice * gas limit + tx.value)
+	overhead, err := s.RollupGpo.SuggestOverhead(context.Background())
+	if err != nil {
+		return err
 	}
 
-	userFee := new(big.Int).Mul(new(big.Int).SetUint64(tx.Gas()), tx.GasPrice())
-	expectedFee := new(big.Int).Mul(expectedTxGasLimit, fees.BigTxGasPrice)
+	raw := new(bytes.Buffer)
+	if err := tx.EncodeRLP(raw); err != nil {
+		return err
+	}
+	// Calculate the fee being paid by the user
+	// Compute the amount of L1 gas that would be used by the transaction
+	l1GasUsed := fees.CalculateL1Cost(raw.Bytes(), overhead.Uint64())
+	// Multiply the L1 gas used by the known L1 gas price to get
+	// the total L1 cost
+	l1Cost := new(big.Int).Mul(l1GasUsed, l1GasPrice)
+	// Multiply tx.gasLimit by tx.gasPrice to get the total L2 cost
+	l2Cost := new(big.Int).Mul(new(big.Int).SetUint64(tx.Gas()), tx.GasPrice())
+	// Add the L1 cost and the L2 cost to get the total fee being paid
+	userFee := new(big.Int).Add(l1Cost, l2Cost)
+
+	// Calculate the expected fee
+	expectedL2Cost := new(big.Int).Mul(new(big.Int).SetUint64(tx.Gas()), l2GasPrice)
+	expectedFee := new(big.Int).Add(expectedL2Cost, l1Cost)
+
```

Comment:
We still need to check that the user has enough balance to pay for the L1 cost + L2 cost. This code exists in `tx_pool.go`